### PR TITLE
Add `String#index/rindex!` methods

### PR DESCRIPTION
**Pull request by Sija**
Adds `String#index!` and `String#index!` methods, similar to the ones found in `Enumerable`.

Supersedes/closes #12729

### CODE REVIEW

File: src/string.cr [hunk 1]
```diff
@@ -3388,6 +3395,27 @@ class String
     match_result.try &.begin
   end
 
+  # :ditto:
+  #
+  # Raises `IndexError` if there is no *search* occurence in `self`.
```
Comment (straight-shoota) [inline src/string.cr:3400]: This might sound better?
```suggestion
  # Raises `IndexError` if *search* does not occur in `self`.
```

**Review by straight-shoota — APPROVED** [review]

**Comment by asterite** [comment]
IndexError doesn't make sense here. Index error means the index was out of bounds. That's not the case here.

**Comment by Sija** [comment]
@asterite Any ideas for a better ones?

**Comment by Sija** [comment]
@asterite I have a bit of an issue with this, since `String` is not an `Enumerable`, technically speaking.

**Comment by asterite** [comment]
Then `NotFoundError` should be moved to the top level.

**Comment by asterite** [comment]
Just like `IndexError`, `KeyError`, etc., are all at the top level. We can probably keep the ones in Enumerable as an alias of the top level ones.

Just my opinion though! I'm sure others will propose more complex solutions.

**Comment by Sija** [comment]
@asterite `NotFoundError` in the top-level is gonna be IMO too generic.

**Comment by straight-shoota** [comment]
Let's just use `Enumerable::NotFoundError` here as we do in `Enumerable#index!`.
Move that type to the top-level namespace should be a separate discussion.

**Comment by asterite** [comment]
Why `NotFoundError` would be too generic, but `IndexError` or `KeyError` would not? Maybe it could be called `ElementNotFoundElement` or something like that.

Another idea: just use whatever `.not_nil!` does for now.

File: src/string.cr [hunk 2]
```diff
@@ -3282,6 +3282,13 @@ class String
     self.match(search, offset).try &.begin
   end
 
+  # :ditto:
+  #
+  # Raises `IndexError` if *search* does not occur in `self`.
```
Comment (beta-ziliani) [inline src/string.cr:3287]:
```suggestion
  # Raises `Enumerable::NotFoundError` if *search* does not occur in `self`.
```

File: src/string.cr [hunk 3]
```diff
@@ -3388,6 +3395,27 @@ class String
...
+  def rindex!(search : Regex, offset = size) : Int32
+    rindex(search, offset) || raise Enumerable::NotFoundError.new
+  end
+
+  # :ditto:
+  #
+  # Raises `IndexError` if *search* does not occur in `self`.
```
Comment (beta-ziliani) [inline src/string.cr:3407]:
```suggestion
  # Raises `Enumerable::NotFoundError` if *search* does not occur in `self`.
```

**Comment by Sija** [comment]
@beta-ziliani Good catch, fixed!

**Comment by HertzDevil** [comment]
`#rindex!` is uncalled for; #11566 didn't add it to the standard library and Ameba does not warn on it. It would be odd to have it here before the collection types do.

**Comment by straight-shoota** [comment]
IMO `#rindex!` makes as much sense as `#index!`, so there would be no harm in adding it. We should add it to `Enumerable` in the same manner.

**Comment by HertzDevil** [comment]
I fully agree with adding `#rindex!`, I just think it should be organized into:

* `String#index!`
* `#rindex!` on everything

rather than:

* `String#index!` + `String#rindex!`
* `#rindex!` on everything else

**Comment by straight-shoota** [comment]
Oh, okay. "#rindex! is uncalled for;" reads like you wouldn't approve of it whatsoever.

IMO the organizational aspects don't matter much. `String#index!` and `String#rindex!` are related and we have a ready-to-go PR for them. I see little point in moving things around into different PRs.

**Pull request merged**